File: src/main/java/urlshortener/transcoding/Base62Converter.java
package urlshortener.transcoding;

public class Base62Converter {

    private static final String BASE62_ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
    private static final int BASE = BASE62_ALPHABET.length();

    public static String encode(long id) {
        StringBuilder encoded = new StringBuilder();
        do {
            encoded.append(BASE62_ALPHABET.charAt((int) (id % BASE)));
            id /= BASE;
        } while (id > 0);
        return encoded.reverse().toString();
    }

    public static long decode(String base62) {
        long decoded = 0;
        for (char c : base62.toCharArray()) {
            int index = BASE62_ALPHABET.indexOf(c);
            if (index == -1) {
                throw new IllegalArgumentException("Invalid character in Base62 string: " + c);
            }
            decoded = decoded * BASE + index;
        }
        return decoded;
    }
}

File: transcoding/Base62Converter.go
package transcoding

import "fmt"

const Alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
const Base = int64(len(Alphabet))

func EncodeBase62(id int64) string {
	encoded := ""
	for id > 0 {
		remainder := id % Base
		encoded = string(Alphabet[remainder]) + encoded
		id /= Base
	}

	return encoded
}

func DecodeBase62(encoded string) (int64, error) {
	decoded := int64(0)
	for _, char := range encoded {
		index := int64(-1)
		for j, alphabetChar := range Alphabet {
			if char == alphabetChar {
				index = int64(j)
				break
			}
		}
		if index == -1 {
			return 0, fmt.Errorf("invalid character in Base62 string")
		}
		decoded = decoded*Base + index
	}

	return decoded, nil
}
